Separate waking the event loop, from communicating with Compositor

gterzian opened the pull request:
<!-- Please describe your changes on the following line: -->

@paulrouget first step of #15934, Glutin only for now, please take a look...

If this makes sense, I will also update the code for ports other than Glutin...

One question: one do I add the `warn!` macro to `servolib`? Also perhaps we would also want to add `box`, since I had to switch to using `Box::new`...

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/17068)
<!-- Reviewable:end -->

gterzian committed:
separate waking the event loop, from communicating with a compositor

jdm commented:
r? @paulrouget

gterzian commented:
UPDATE: also added the implementation for the cef port...

paulrouget commented on ports/glutin/window.rs:
```diff
@@ -1047,17 +1046,27 @@ impl WindowMethods for Window {
-         box receiver as Box<CompositorReceiver>)
+        box GlutinEventLoopRiser {
+            window_proxy: window_proxy,
+        } as Box<EventLoopRiser + Send>
```
Is the `as …` necessary?

paulrouget reviewed (commented):
Looks good to me. There are some nits to address. Also, I'm not 100% sure you need to use Box<CompositorProxy>.

Once these addressed, I'd like @asajeffrey to take a look as well.

paulrouget commented on ports/glutin/window.rs:
```diff
@@ -1047,17 +1046,27 @@ impl WindowMethods for Window {
+            fn clone(&self) -> Box<EventLoopRiser + Send> {
+                box GlutinEventLoopRiser {
+                    window_proxy: self.window_proxy.clone(),
+                } as Box<EventLoopRiser + Send>
```
Is the `as …` necessary?

gterzian commented:
@paulrouget Ok, I replaced the traits with structs, and addressed the other comments too... 

I'm wondering if we need to add some sort of "clone" method to `CompositorReceiver`?

paulrouget commented on ports/cef/window.rs:
```diff
@@ -295,13 +295,18 @@ impl WindowMethods for Window {
-         } as Box<CompositorProxy+Send>,
-         box receiver as Box<CompositorReceiver>)
+    fn create_event_loop_riser(&self) -> Box<EventLoopRiser> {
+        struct CefEventLoopRiser
```
Add a `;` at the end after `CefEventLoopRiser`.

paulrouget commented on ports/cef/window.rs:
```diff
@@ -295,13 +295,18 @@ impl WindowMethods for Window {
+            }
+            fn clone(&self) -> Box<EventLoopRiser + Send> {
+                box CefEventLoopRiser {
+                } as Box<EventLoopRiser + Send>
```
I think you can just do:
```rust
fn clone() {
  box CefEventLoopRiser
}
```

paulrouget commented:
Thank you Greg. This looks good to me now (beside a nit).

I'd like to get a review from Alan.

@asajeffrey - this is the first step to clarify the Embedder <-> Constellation communication. Can you take a look too?

asajeffrey commented on components/compositing/compositor_thread.rs:
```diff
@@ -22,40 +22,54 @@ use style_traits::viewport::ViewportConstraints;
-    fn clone_compositor_proxy(&self) -> Box<CompositorProxy + 'static + Send>;
+
+/// Used to wake up the event loop, provided by the servo port/embedder.
+pub trait EventLoopRiser : 'static + Send {
```
Nit: not sure about the name. A riser is a part of a staircase :)

gterzian replied:
How about `EventLoopAwaker`, so that we have a noun? https://en.wiktionary.org/wiki/awaker

asajeffrey reviewed (approved):
LGTM, I just have an annoying bikeshedding name change. `EventLoopWakeup`?

paulrouget commented:
```
ajeffrey> "Waker" would probably be more colloquial English: "Awaker" sounds like "I am more awake than you".
```

gterzian commented:
Ok, `EventLoopWaker` it is...

gterzian commented:
It looks like approving in Github isn't enough, and somebody needs to do a r+ at the bot...

asajeffrey commented:
@bors-servo r+

bors-servo commented:
:clipboard: Looks like this PR is still in progress, ignoring approval

asajeffrey commented:
Hmm, I wonder why bors reckons this is still a WIP. Is it the unchecked boxes in the PR description?

paulrouget commented:
Probably the title. I changed it. Do you need to re-approve?

paulrouget commented:
https://github.com/bundler/homu/blob/288bda90d87f6500f7dd9dd779b633f7615f72e4/homu/main.py#L352

asajeffrey commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 3a693c7 has been approved by `asajeffrey`

<!-- @bors-servo r=asajeffrey 3a693c7a23e7c3cc1d3e57f12d1eff4b83152134 -->

gterzian commented:
@paulrouget @asajeffrey thanks for the review.

bors-servo commented:
:hourglass: Testing commit 3a693c7a23e7c3cc1d3e57f12d1eff4b83152134 with merge 7e273d6c9b86d6ffbf216e84ae7326976888e5ef...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/7366), [arm32](http://build.servo.org/builders/arm32/builds/7580), [arm64](http://build.servo.org/builders/arm64/builds/7619), [linux-dev](http://build.servo.org/builders/linux-dev/builds/7708), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/4464), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/4488), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/7691), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/175), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/174), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/4394), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/4458), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/174), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/177), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/2935)
Approved by: asajeffrey
Pushing 7e273d6c9b86d6ffbf216e84ae7326976888e5ef to master...

paulrouget commented:
Thank you @gterzian

asajeffrey commented:
@gterzian thanks!
